### PR TITLE
Add onRequestClose prop to Modal to ensure state is driven purely by consuming components

### PR DESCRIPTION
**Pull request by Matt-Butler**
### Summary
For dismissing the modal with clicking on the overlay or pressing the esc key, we were using react-portal. This was not updating our isOpened state. This PR is correcting the value of the isOpened state.

https://github.com/cerner/terra-core/issues/293

After these changes, the isOpened state is accurate.
isopened_state

Thanks for contributing to Terra. 
@cerner/terra

### CODE REVIEW

File: packages/terra-modal/src/Modal.jsx [hunk 1]
```diff
@@ -79,8 +83,28 @@ const defaultProps = {
...
+
 class Modal extends React.Component {
+  constructor() {
+    super();
+    this.handleKeydown = this.handleKeydown.bind(this);
+  }
+
+  componentDidMount() {
+    if (this.props.closeOnEsc) {
+      document.addEventListener('keydown', this.handleKeydown);
```
Comment (tbiethman) [inline packages/terra-modal/src/Modal.jsx]: We probably want to remove this on componentWillUnmount

File: packages/terra-modal/src/Modal.jsx [hunk 2]
```diff
@@ -97,13 +121,14 @@ class Modal extends React.Component {
...
           role,
+          onRequestClose,
            ...customProps } = this.props;
 
     return (
       <Portal
         isOpened={isOpened}
-        closeOnEsc={closeOnEsc}
-        closeOnOutsideClick={closeOnOutsideClick}
+        closeOnEsc={false}
```
Comment (tbiethman) [inline packages/terra-modal/src/Modal.jsx]: Could we just remove these guys completely? Or do they default the values to true?

File: packages/terra-modal/src/Modal.jsx (hunk 1 above)
Comment (Matt-Butler) [inline packages/terra-modal/src/Modal.jsx]: +1

File: packages/terra-modal/src/Modal.jsx (hunk 2 above)
Comment (Matt-Butler) [inline packages/terra-modal/src/Modal.jsx]: I'll test that right now. react-portal does not provide default props for either of them.

File: packages/terra-modal/src/Modal.jsx [hunk 3]
```diff
@@ -79,8 +83,28 @@ const defaultProps = {
...
+};
+
 class Modal extends React.Component {
+  constructor() {
+    super();
+    this.handleKeydown = this.handleKeydown.bind(this);
+  }
+
+  componentDidMount() {
+    if (this.props.closeOnEsc) {
```
Comment (tbiethman) [inline packages/terra-modal/src/Modal.jsx]: If, for whatever reason, someone passed the closeOnEsc prop to the Modal component after its initial mounting, we wouldn't add the event handler.

I'm not sure how realistic a scenario that is, but we could always add the event handler and check the prop in the  handler to cover that case.

Comment (Matt-Butler) [inline packages/terra-modal/src/Modal.jsx]: So you recommend always adding the event handler and only trigger onRequestClose when the prop is set?

File: packages/terra-modal/src/Modal.jsx (hunk 1 above)
Comment (Matt-Butler) [inline packages/terra-modal/src/Modal.jsx]: https://github.com/cerner/terra-core/pull/321/commits/c171e322755b7e4616000fc0b6715820927dd9d0

File: packages/terra-modal/src/Modal.jsx (hunk 3 above)
Comment (tbiethman) [inline packages/terra-modal/src/Modal.jsx]: I think so. It'd be easier than checking the props on didUpdate every time.

File: packages/terra-modal/src/Modal.jsx (hunk 2 above)
Comment (Matt-Butler) [inline packages/terra-modal/src/Modal.jsx]: https://github.com/cerner/terra-core/pull/321/commits/de085f27a1f398ca35f224f1e84353e61cb69c4d
Remove the props set to false. We get the same behavior without them set.

File: packages/terra-modal/src/Modal.jsx (hunk 3 above)
Comment (Matt-Butler) [inline packages/terra-modal/src/Modal.jsx]: https://github.com/cerner/terra-core/pull/321/commits/a0e51cb62cb98405696bb575141f7b7f8bb12a37

**Pull request merged**